scipy/optimize/_qap.py

```diff
@@ -0,0 +1,722 @@
+    elif init_k <= 0:
+        msg = "'init_k' must be a positive integer"
+    elif maxiter <= 0:
+        msg = "'maxiter' must be a positive integer"
+    elif eps <= 0:
+        msg = "'eps' must be a positive float"
+    if msg is not None:
+        raise ValueError(msg)
+
+    # TypeError check
+    if not isinstance(shuffle_input, bool):
+        msg = "'shuffle_input' must be a boolean"
+    if msg is not None:
+        raise TypeError(msg)
+
+    rng = check_random_state(rng)
+    n = A.shape[0]  # number of vertices in graphs
+    n_seeds = partial_match.shape[0]  # number of seeds
+    n_unseed = n - n_seeds
+
+    perm_inds = np.zeros(n)
+
+    obj_func_scalar = 1
+    if maximize:
+        obj_func_scalar = -1
+    score = obj_func_scalar * np.inf
+
+    seed_dist_c = np.setdiff1d(range(n), partial_match[:, 1])
+    if shuffle_input:
+        seed_dist_c = rng.permutation(seed_dist_c)
+        # shuffle_input to avoid results from inputs that were already matched
+
+    seed_cost_c = np.setdiff1d(range(n), partial_match[:, 0])
+    permutation_cost = np.concatenate([partial_match[:, 0],
+                                       seed_cost_c], axis=None).astype(int)
+    permutation_dist = np.concatenate([partial_match[:, 1],
+                                       seed_dist_c], axis=None).astype(int)
+    A = A[np.ix_(permutation_cost, permutation_cost)]
+    B = B[np.ix_(permutation_dist, permutation_dist)]
+
+    # definitions according to Seeded Graph Matching [2].
+    A11 = A[:n_seeds, :n_seeds]
+    A12 = A[:n_seeds, n_seeds:]
+    A21 = A[n_seeds:, :n_seeds]
+    A22 = A[n_seeds:, n_seeds:]
+    B11 = B[:n_seeds, :n_seeds]
+    B12 = B[:n_seeds, n_seeds:]
+    B21 = B[n_seeds:, :n_seeds]
+    B22 = B[n_seeds:, n_seeds:]
+
+    # setting initialization matrix
+    if isinstance(init_J, str) and init_J == 'barycenter':
+        J = np.ones((n_unseed, n_unseed)) / float(n_unseed)
+    else:
+        _check_init_input(init_J, n_unseed)
+        J = init_J
+
+    total_iter = 0
+    for i in range(init_k):
+        if init_weight is not None:
+            # generate a nxn matrix where each entry is a random number [0, 1]
+            K = rng.random((n_unseed, n_unseed))
+            # Sinkhorn balancing
+            K = _doubly_stochastic(K)
+            # initialize J, a doubly stochastic barycenter
+            P = J * init_weight + (1 - init_weight) * K
+        else:
+            P = J
+        const_sum = A21 @ B21.T + A12.T @ B12
+        grad_P = np.inf  # gradient of P
+        n_iter = 0  # number of FW iterations
+
+        # OPTIMIZATION WHILE LOOP BEGINS
+        while grad_P > eps and n_iter < maxiter:
+            # computing the gradient of f(P) = -tr(APB^tP^t)
+            delta_f = (const_sum + A22 @ P @ B22.T + A22.T @ P @ B22)
+            # run hungarian algorithm on gradient(f(P))
+            rows, cols = linear_sum_assignment(obj_func_scalar * delta_f)
```

Comment:
`rows, cols`
These variables do not need to be unpacked.

Reply:
I'd prefer to leave that alone, if that's OK? Unpacking is not one of the more expensive operations here, and I think it's readable.